packages/dds/tree/src/core/tree/visithPath.ts

```diff
@@ -0,0 +1,34 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import * as Delta from "./delta";
+import { UpPath } from "./pathTree";
+import { Value } from "./types";
+
+/**
+ * Delta visitor for the path tree.
+ *
+ * TODO: additional callbacks
+ * @alpha
+ */
+export interface PathVisitor {
+	/**
+	 * A sequence of nodes of length `count` is being deleted starting with `path`.
+	 * Called when these nodes are no longer parented under their previous parent, and do not have a new parent.
+	 * It is possible they may be un-deleted in the future (for example by a conflicted merge or undo).
+	 *
+	 * Not called for children of deleted nodes.
+	 *
+	 * @param path - first node in the deleted range.
+	 * @param count - length of deleted range.
+	 */
+	onDelete(path: UpPath, count: number): void;
+	/**
+	 * @param path - location which first node of inserted range will have after insert.
+	 * Any nodes at this index (or after it) will be moved to the right (have their indexes increased by `content.length`).
+	 * @param content - content which is being inserted.
+	 */
+	onInsert(path: UpPath, content: Delta.ProtoNodes): void;
+	onSetValue(path: UpPath, value: Value): void;
```

Comment:
it is invalid to access the contents of the anchorset or tree at the time these events occur, so it doesn't matter, or more specifically for any code where it matters in invalid code.

Note that onDelete happens when stuff is perma dead (can not be un-deleted) not when the content is initially deleted, so thats a much different form of after than when onInsert happens relative to when someone actually does the insert (which you can't really observe the ordering of other than that the indexes used for future edits to the filed use the index numbers as if the insert had already happened).

Related: exactly when things actually get deleted instead of just moved to a detached location is a work in progress as that area is going through design changes and still being implemented. Since we don't have proper move events yet, events for things getting moved out as the first stage of delete also don't exist.